signer/xpi/xpi.go

```diff
@@ -142,20 +154,145 @@ func (s *PKCS7Signer) Config() signer.Configuration {
 	}
 }
 
-// SignFile takes an unsigned zipped XPI file and returned a signed XPI file
-func (s *PKCS7Signer) SignFile(input []byte, options interface{}) (signer.SignedFile, error) {
+// isCOSEAlgBad checks whether a COSE algorithm is recognized as a
+// valid IANA algorithm name, and supported and enabled in autograph
+func isCOSEAlgBad(algName string) (alg *cose.Algorithm, err error) {
+	alg, algErr := cose.GetAlgByName(algName)
+	if algErr != nil {
+		err = errors.Wrapf(algErr, "xpi: unrecognized algorithm COSE Signing algorithm")
+		return
+	}
+	if _, exists := supportedCOSEAlgNames[alg.Name]; exists == false {
+		err = fmt.Errorf("xpi: COSE algorithm %s is not supported", algName)
+		return
+	}
+	return alg, nil
+}
+
+// SignFile takes an unsigned zipped XPI file and returns a signed XPI file
+func (s *PKCS7Signer) SignFile(input []byte, options interface{}) (signedFile signer.SignedFile, err error) {
```

Comment:
I feel like a lot of this code could be refactored into a separate function to be easier to follow. Unsure how, but the COSE part is much longer than the PKCS7 one for some reason, which feels unbalanced. Thoughts?

Reply:
I'd like to do some of the following to simplify it:

* update the go-cose library (need to do this anyway since it was using math/rand instead of crypto/rand)
  * don't use signer and verifier opts
  * compress headers before signing
  * pass signature params to the message sign function so we don't have to construct and add them
  * I have WIP to have `NewSignMessage` return a pointer, ran into some errors though
* move COSE signing config to signer options
  * this lets us remove the `isCOSEAlgBad` / build `coseSigAlgs` loop
* defined types for a manifest and JAR/XPI that let us add files one by one and not read the ZIP add them in SignFile

There will probably be more inherent complexity in generating multiple EE certs and signatures, but it'd be cleaner to move that logic out of the function.